File: src/__tests__/clients/stress_ui/y_websocket_client_insert.mjs
import WebSocket from "ws";
import { Doc } from 'yjs';
import { WebsocketProvider } from 'y-websocket';

const id = Number(process.argv[2])
const textLength = Number(process.argv[3])
const warmupPeriodSeconds = Number(process.argv[4])

const doc = new Doc();
const t = doc.getText('t');


let MUTATE_DOC = true;


let wsProvider = new WebsocketProvider(
  'ws://127.0.01:8888/jupyter_rtc_test/room',
  'jupyter_rtc_test',
  doc,
  { WebSocketPolyfill: WebSocket }
);

function randomString(length) {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const charactersLength = characters.length;
  let counter = 0;
  while (counter < length) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
    counter += 1;
  }
  return result;
}

const infoWebSocket = new WebSocket('ws://127.0.0.1:8888/jupyter_rtc_test/stresser');
infoWebSocket.onopen = () => {
  const info = { id, action: 'info', text: t.toString() }
  infoWebSocket.send(JSON.stringify(info));
};
infoWebSocket.onmessage = (message) => {
  const data = JSON.parse(message.data);
  if (data.action === 'pause') {
    MUTATE_DOC = false;
  }
  if (data.action === 'restart') {
    MUTATE_DOC = true;
  }
};

wsProvider.on('status', event => {
//  console.log('Status', event);
  if (event.status === 'connected') {
    setInterval(() => {
      if (MUTATE_DOC) {
        if (t.length < textLength) {
          t.insert(0, randomString(4));
        } else {  
          t.delete(0, t.length / 2);
        }
        console.log('Nodejs client', id, t.toString());
      }
    }, 1000);
  }
});

wsProvider.on('sync', isSynced => {
  console.log('Nodejs client isSynced', id, t.toString());
});
/*
t.observe(event => {
  console.log('event', t.toString());
});
*/

File: jupyter_rtc_test/stresser/clients/y_websocket_client_insert.py
import random
import json
import string
import sys
import rel

import websocket
from websocket import WebSocket

import asyncio

from websockets import connect  # type: ignore

from y_py import YDoc
from ypy_websocket import WebsocketProvider


id = int(sys.argv[1])
text_length = int(sys.argv[2])
warmup_period_seconds = int(sys.argv[3])

doc = YDoc()
text = doc.get_text('t')


MUTATE_DOC = True


info_ws_client = WebSocket()
info_ws_client.connect("ws://127.0.01:8888/jupyter_rtc_test/stresser")


def on_message(ws, message):
    print("-----------------", message)

info_ws_client_listener = websocket.WebSocketApp("ws://127.0.01:8888/jupyter_rtc_test/stresser", on_message=on_message)
info_ws_client_listener.run_forever(dispatcher=rel, reconnect=5)

async def main():
    websocket = await connect("ws://127.0.01:8888/jupyter_rtc_test/room/jupyter_rtc_test")
    websocket_provider = WebsocketProvider(doc, websocket)
    while True:
        if MUTATE_DOC:
            with doc.begin_transaction() as txn:
                length = len(str(text))
                if length < text_length:
                    random_string = ''.join(random.choices(string.ascii_uppercase + string.digits, k=3))
                    text.insert(txn, random.randint(0, length), random_string)
                else:
                    text.delete(txn, random.randint(0, int(length / 2)))
            print("Python client text", id, str(text))
            payload = json.dumps({
                "id": id,
                "action": "info",
                "text": str(text),
            })
            info_ws_client.send(payload)
        await asyncio.sleep(1)


if __name__ == '__main__':
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass
    loop.close()
